src/evaluators/gather/PHAL_GatherScalarNodalParameter_Def.hpp

```diff
@@ -320,4 +320,156 @@ evaluateFields(typename Traits::EvalData workset)
+{
+  this->setName("GatherNodalParameter("+this->param_name+")"+PHX::print<PHAL::AlbanyTraits::HessianVec>());
+}
+
+// **********************************************************************
+template<typename Traits>
+GatherScalarNodalParameter<PHAL::AlbanyTraits::HessianVec, Traits>::
+GatherScalarNodalParameter(const Teuchos::ParameterList& p) :
+  GatherScalarNodalParameterBase<PHAL::AlbanyTraits::HessianVec, Traits>(p,p.get<Teuchos::RCP<Albany::Layouts> >("Layouts Struct"))
+{
+  this->setName("GatherNodalParameter("+this->param_name+")"+PHX::print<PHAL::AlbanyTraits::HessianVec>());
+}
+
+// **********************************************************************
+template<typename Traits>
+void GatherScalarNodalParameter<PHAL::AlbanyTraits::HessianVec, Traits>::
+evaluateFields(typename Traits::EvalData workset)
+{
+  if (this->memoizer.have_saved_data(workset,this->evaluatedFields())) return;
+
+  // Distributed parameter vector
+  Teuchos::RCP<const Thyra_Vector> pvec = workset.distParamLib->get(this->param_name)->overlapped_vector();
+  Teuchos::ArrayRCP<const ST> pvec_constView = Albany::getLocalData(pvec);
+
+  // Direction vector for the Hessian-vector product
+  Teuchos::RCP<const Thyra_MultiVector> vvec = workset.hessianWorkset.direction_p;
+
+  auto nodeID = workset.wsElNodeEqID;
+  const Albany::IDArray& wsElDofs = workset.distParamLib->get(this->param_name)->workset_elem_dofs()[workset.wsIndex];
+
+  bool g_xp_is_active = !workset.hessianWorkset.hess_vec_prod_g_xp.is_null();
+  bool g_px_is_active = !workset.hessianWorkset.hess_vec_prod_g_px.is_null();
+  bool g_pp_is_active = !workset.hessianWorkset.hess_vec_prod_g_pp.is_null();
+
+  // Are we differentiating w.r.t. this parameter?
+  bool is_active = (workset.dist_param_deriv_name == this->param_name);
+  bool is_direction_active = (workset.hessianWorkset.dist_param_deriv_direction_name == this->param_name);
+
+  Teuchos::ArrayRCP<const ST> vvec_constView;
+  if(is_direction_active && (g_xp_is_active || g_pp_is_active)) {
+    TEUCHOS_TEST_FOR_EXCEPTION(
+        vvec.is_null(),
+        Teuchos::Exceptions::InvalidParameter,
+        "\nError in GatherScalarNodalParameter<HessianVec, Traits>: "
+        "direction_p is not set and hess_vec_prod_g_xp or "
+        "hess_vec_prod_g_pp is set.\n");
+    vvec_constView = Albany::getLocalData(vvec->col(0));
+  }
+
+  const int num_nodes = this->numNodes;
+
+  const int num_deriv = (this->val)(0,0).size();
+  for (std::size_t cell=0; cell < workset.numCells; ++cell ) {
+    for (std::size_t node = 0; node < num_nodes; ++node) {
+
+      // Initialize Fad type for parameter value
+      const LO id = wsElDofs((int)cell,(int)node,0);
+      RealType pvec_val = (id >= 0) ? pvec_constView[id] : 0;
+
+      auto val = (this->val)(cell,node);
+      val = FadType(num_deriv, pvec_val);
+      if (is_direction_active && (g_xp_is_active||g_pp_is_active))
+        val.val().fastAccessDx(0) = (id >= 0) ? vvec_constView[id] : 0;
```

Comment:
Why do we modify only Dx(0)? Could you add a quick comment on what's happening here? I would have expected something like `Dx(param_index)`

Reply:
The reason is that, we are using the <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{H}_{\boldsymbol{x}\boldsymbol{p}_1}\boldsymbol{v}_{\boldsymbol{p}_1}=\left.\frac{\partial}{\partial r} \nabla_{\boldsymbol{x}} g(\boldsymbol{x},\boldsymbol{p}_1%2Br \boldsymbol{v}_{\boldsymbol{p}_1}, \boldsymbol{p}_2)\right|_{r=0}"> formula where <img src="https://render.githubusercontent.com/render/math?math=\large r"> is a scalar to compute the Hessian-vector product contributions. 

Therefore, we have two nested derivatives and the outermost one (mathematically), i.e. the one with respect to <img src="https://render.githubusercontent.com/render/math?math=\large r">, is the one related to the innermost `FAD` type. We know that this innermost `FAD` type will always have a size of one and only `Dx(0)` has to be specified. It is clearer?

Reply:
Oh, the inner most Fad does not have a length equal to the number parameters, but rather to the current parameter "vector dimension"?

Reply:
Oh, I think that I have misunderstood what you meant, especially when reading `Dx(param_index)`.
Here, up to now, we only evaluate one contribution of the Hessian-vector product at a time when we are in that `evaluateFields` function. This means that, somehow, we have basically only one current parameter "vector dimension", yes.

Later, we might consider evaluate multiple contributions together, it will increase the sizes of the `FAD` types but it could be more cost effective.

Reply:
Would you mind adding a comment in the code explaining what you are doing? I know that there are very little comments in Albany but we are trying to add more comments as we go. Maybe there are other places where a few words of comments would hlep understanding what's implemented. Thanks!